### PR TITLE
[enhancement](Nereids) support decimalv3 type

### LINKED ISSUE #17500: support decimal v3 type in nereids

### LINKED ISSUE #17501: let decimal v2 and datetimev2 precision infer same with legacy one

### PR DESCRIPTION
**Pull request by morrySnow**
# Proposed changes

Issue Number: close #17500 close #17501

1. support decimalv3 type
2. fix all type coercion different with legacy planner
3. support datetimev2 precision derivation
4. fix decimalv2 precision derivation error
5. arithmetic expr generate function by itself when do expr translation
6. cast(x as char(n)) will turn into substr(cast(x as char), 1, n) in parser
7. remove useless finalizeImplForNereids code in FunctionCallExpr

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morrySnow** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.56 seconds
 stream load tsv:          465 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         42 seconds loaded 2358488459 Bytes, about 53 MB/s
 stream load orc:          75 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230303115454_clickbench_pr_108072.html

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run p0

**Comment by morrySnow** [comment]
run p0

**Comment by morrySnow** [comment]
run compile

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 1]
```diff
@@ -750,15 +758,7 @@ protected void compactForLiteral(Type type) throws AnalysisException {
 
     @Override
     public void finalizeImplForNereids() throws AnalysisException {
-        if (op == Operator.BITNOT) {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:753]: do not need finalizeImplForNereids?
how to init fn?

**Review by englefly — APPROVED** [review]

**Review by jackwener — APPROVED** [review]

**Pull request merged**